File: 2022/day-3/RucksackReorganization.java
import java.io.File;
import java.util.HashMap;
import java.util.Map;
import java.util.Scanner;
import java.util.Map.Entry;

public class RucksackReorganization {
    public static void main(String[] args) throws Exception {
        partOne("/home/ultron/Documents/Projects/advent-of-code/2022/day-3/sample.txt");
    }

    public static void partOne(String filename) throws Exception {
        String commonCharacter = "v";
        Scanner scanner = new Scanner(new File(filename));
        char currentCharacter;
        int total = 0;

        while (scanner.hasNextLine()) {
            String currentLine = scanner.nextLine();
            Map<Character, Integer> itemCount = new HashMap<>();
            char doubleItem = 'x';

            for (int i = 0; i < currentLine.length() / 2; i++) {
                itemCount.put(currentLine.charAt(i), 1);
            }

            for (int i = currentLine.length() / 2; i < currentLine.length(); i++) {
                currentCharacter = currentLine.charAt(i);
                if (itemCount.containsKey(currentCharacter)) {
                    itemCount.put(currentCharacter, itemCount.get(currentCharacter) + 1);
                } else {
                    itemCount.put(currentLine.charAt(i), 1);
                }
            }
            System.out.println(itemCount);

            for (Entry<Character, Integer> entry : itemCount.entrySet()) {
                doubleItem = entry.getKey();
                if (entry.getValue() == 2) {
                    if ((int) Character.valueOf(doubleItem) >= 65 & (int) Character.valueOf(doubleItem) <= 65)
                        total += (int) Character.valueOf(entry.getKey()) % 65 + 27;
                    if ((int) Character.valueOf(doubleItem) >= 97 & (int) Character.valueOf(doubleItem) <= 65)
                        total += (int) Character.valueOf(entry.getKey()) % 96;
                }
            }
        }

        System.out.println(total);
    }
}

File: 2022/day-3/rucksack_reorganization.py
def rucksack_reorganization_one():
    lines = open("/home/ultron/Documents/Projects/advent-of-code/2022/day-3/input.txt", "r").readlines()
    
    total = 0
    
    for line in lines:
        left = set(line[:int(len(line) / 2)])
        right = set(line[int(len(line) / 2):])
        
        common = left.intersection(right).pop()
        if 97 <= ord(common) <= 122:
            total += ord(common) % 96
        elif 65 <= ord(common) <= 90:
            total += ord(common) % 64 + 26

    return total

def rucksack_reorganization_two():
    lines = open("/home/ultron/Documents/Projects/advent-of-code/2022/day-3/input.txt", "r").readlines()

    total = 0

    for i in range(0,len(lines),3):
        member1 = set(lines[i].strip())
        member2 = set(lines[i + 1].strip())
        member3 = set(lines[i + 2].strip())

        common = member1.intersection(member2).intersection(member3).pop()
        if 97 <= ord(common) <= 122:
            total += ord(common) % 96
        elif 65 <= ord(common) <= 90:
            total += ord(common) % 64 + 26
        
    return total


if __name__ == "__main__":
    # print(rucksack_reorganization_one())
    print(rucksack_reorganization_two())